### PR TITLE
Implement flexible box layout for row container

### PR DESCRIPTION
**Pull request by stshine**
This pull requests implements basic flexible box layout for row container.
It  has implemented most basic flexbox features, including grow, shrink, multi-line, *reverse properties, and alignment under `justify-content`, `align-items`, `align-self`, `align-content`.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X]  There are tests for these changes

r? @pcwalton

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12330)

### CODE REVIEW

**Comment by stshine** [comment]
There a several concerns for this pull request.

One is that previously every method involves a vtable lookup, so I add a style field and clone the style of the flow to it, but I am not very sure what the price is.
Another one is that there are occurrences of 'box-sizing' property here and there to get the proper main size. I wonder if there is a way to reduce them.

**Comment by stshine** [comment]
ping @pcwalton 
Mind share some general opinion about this pull request? There is a bunch of PRs pending on this.

**Comment by shinglyu** [comment]
(tracking #12453)

**Comment by shinglyu** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/layout/flex.rs, line 145 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMwN6VAs86npoGM6RhZ:-KMwN6VAs86npoGM6Rh_:1563241048) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L145)):*
> ```Rust
> impl FlexItem {
>     pub fn new(flow: FlowRef) -> FlexItem {
>         let style = flow.as_block().fragment.style.clone();
> ```

Will the style ever change after we clone it (e.g. by JavaScript)? Do we need to sync it up with the style side?

---

*[components/layout/flex.rs, line 222 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMwOZ5Jp9XAoGomfnEJ:-KMwOZ5Jp9XAoGomfnEK:1993642271) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L222)):*
> ```Rust
>         let ref fragment = self.flow.as_block().fragment;
>         let adjustment;
>         match mode {
> ```

We could just `let adjustment = match mode {...}`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12330)*

**Comment by shinglyu** [comment]
Disclaimer: I don't have the right to r+ yet, just providing some feedback.

**Comment by stshine** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/layout/flex.rs, line 145 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMwN6VAs86npoGM6RhZ:-KMx3Q5LSHlaiyp-KMyX:-682205792) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L145)):*
<details><summary><i>Previously, shinglyu (Shing Lyu) wrote…</i></summary>
> Will the style ever change after we clone it (e.g. by JavaScript)? Do we need to sync it up with the style side?
</details>
What cloned is just a Arc pointer to the actual `ServoComputedValues`, so It should not be a problem.

---

*[components/layout/flex.rs, line 191 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMxm5sNISQk1u8SsF9t:-KMxm5sO9s9ZF1V_sqNv:50195832) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L191)):*
> ```Rust
>                         (MaybeAuto::from_style(margin.inline_start, Au(0)).specified_or_zero() +
>                          MaybeAuto::from_style(margin.inline_end, Au(0)).specified_or_zero())
>                     }
> ```

Here when `box-sizing` is set to `border-box`, The border and padding length contained in main size is computed at intrinsic size calculating, which set percentage padding to zero. Maybe I should take percentage padding into account for inline main size.

---

*[components/layout/flex.rs, line 222 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMwOZ5Jp9XAoGomfnEJ:-KMx3ieTeOlxGQPXxyup:2091736419) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L222)):*
<details><summary><i>Previously, shinglyu (Shing Lyu) wrote…</i></summary>
> We could just `let adjustment = match mode {...}`
</details>
Previously I do the border and margin computing here. Will fix it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12330)*

**Comment by stshine** [comment]
> Disclaimer: I don't have the right to r+ yet, just providing some feedback.
May I ask your general opinion about this pull request? If you think this pull request is acceptable, I will make other PRs that depend on this.

File: components/layout/flex.rs [hunk 1]
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
...
+    /// Style of the child flow, stored here to reduce overhead.
+    pub style: Arc<ServoComputedValues>,
+    /// The 'flex-grow' property of this item.
+    pub flex_grow: f32,
+    /// The 'flex-shrink' property of this item.
+    pub flex_shrink: f32,
+    /// The 'order' property of this item.
+    pub order: i32,
+    /// Whether the main size has met its constraint.
+    pub is_freezed: bool,
```
Comment (pcwalton) [inline components/layout/flex.rs]: nit: should this be `is_frozen`?

File: components/layout/flex.rs [hunk 2]
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
...
+                self.base_size = basis.specified_or_default(content_size);
+                self.max_size = specified_or_none(self.style.max_block_size(), containing_length)
+                    .unwrap_or(MAX_AU);
+                self.min_size = specified(self.style.min_block_size(), containing_length);
+            }
+        }
+    }
+
+    /// Return the outer main size of the item, including paddings and margins,
+    /// clamped by max and min size.
```
Comment (pcwalton) [inline components/layout/flex.rs]: Maybe this should become a method on `Fragment`?

File: components/layout/flex.rs [hunk 3]
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
...
+    pub fn outer_main_size(&self, mode: Mode) -> Au {
+        let ref fragment = self.flow.as_block().fragment;
+        let adjustment;
+        match mode {
```
Comment (pcwalton) [inline components/layout/flex.rs]: nit: you can shorten this to `let adjustment = match mode { ... }` and then remove the `adjustment =`s below and the semicolons

File: components/layout/flex.rs [hunk 4]
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
...
+            if let LengthOrPercentageOrAuto::Length(length) = main_length {
+                MaybeAuto::Specified(length)
+            } else {
+                MaybeAuto::Auto
+            }
+        }
+    }
+}
+
+/// Represent a child in a flex container. Most fields here are used in
```
Comment (pcwalton) [inline components/layout/flex.rs]: nit: "Represent" -> "Represents"

**Comment by pcwalton** [comment]
Looks good, just had a few nits. Thanks!

File: components/layout/flex.rs [hunk 5]
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
...
+                                                          border_collapse::T::separate);
+                block.fragment.compute_inline_direction_margins(containing_length);
+                block.fragment.compute_block_direction_margins(containing_length);
+
+                let adjust_size = match self.style.get_position().box_sizing {
+                    box_sizing::T::border_box => {
+                        let margin = self.style.logical_margin();
+                        (MaybeAuto::from_style(margin.inline_start, Au(0)).specified_or_zero() +
+                         MaybeAuto::from_style(margin.inline_end, Au(0)).specified_or_zero())
+                    }
```
Comment (stshine) [inline components/layout/flex.rs]: Here when the the `box-sizing` property is `border-box`, percentage padding included in main size is calculated to zero as a result of `bubble_inline_sizes`. Maybe I should recalculate them?

File: components/layout/flex.rs (hunk 2 above)
Comment (stshine) [inline components/layout/flex.rs]: All these methods exists as the result of refactoring the method of obtaining a line in flexbox. Both this and the adjustment issues remain here because I previously do margin and padding calculation in `outer_main_size()` . This point is quite valid, but 1. I am not sure if other components need this method, 2. This will make the style field unnecessary and make `auto_margin_num()` a function instead of a method. Maybe I should do this after more consideration,  together with renaming the `Mode` enum to `Direction` and moving it to `model.rs`.

**Comment by stshine** [comment]
Issues addressed. Thanks!

**Comment by shinglyu** [comment]
Does any other code use these methods? If it's for future patches, maybe we should merge this right before the one that uses it.

**Comment by stshine** [comment]
@shinglyu told me for a big feature I should make separate commits and submit them in on PR. If so I will close other PRs, rename this PR, and push consequent commits for flexbox here.

**Comment by jdm** [comment]
That sounds like a good plan.

**Comment by stshine** [comment]
So this PR is ready. Through it has implement most basic features, There are still a lot of things missing . except the most import support for column container,  some missing features I gathered is:
- `bubble_inline_sizes()` is not implemented correctly;
- absolutely positioned children is not layouted yet;
- `visibility: hidden` is not implemented;
- `align-self: baseline` is not implemented;
- The interaction between block flow and flex flow needs huge enhancement.

However, IMO the core logic is stable enough for review. Since I am not  experienced, criticisms are very much appreciated!

File: components/layout/flex.rs [hunk 6]
```diff
@@ -69,22 +72,192 @@ impl AxisSize {
...
+                    box_sizing::T::border_box =>
+                        block.fragment.border_padding.inline_start_end()
+                };
+                let content_size = block.base.intrinsic_inline_sizes.preferred_inline_size
+                    - block.fragment.surrounding_intrinsic_inline_size() + adjustment;
+                self.base_size = basis.specified_or_default(content_size);
+                self.max_size = specified_or_none(self.style.max_inline_size(), containing_length)
+                    .unwrap_or(MAX_AU);
+                self.min_size = specified(self.style.min_inline_size(), containing_length);
+            },
```
Comment (pcwalton) [inline components/layout/flex.rs]: uber-nit: no need for this `,`

File: components/layout/flex.rs [hunk 7]
```diff
@@ -69,22 +72,192 @@ impl AxisSize {
...
+                block.fragment.compute_inline_direction_margins(containing_length);
+                block.fragment.compute_block_direction_margins(containing_length);
+
+                let adjustment = match self.style.get_position().box_sizing {
+                    box_sizing::T::content_box => Au(0),
+                    box_sizing::T::border_box =>
+                        block.fragment.border_padding.inline_start_end()
+                };
+                let content_size = block.base.intrinsic_inline_sizes.preferred_inline_size
+                    - block.fragment.surrounding_intrinsic_inline_size() + adjustment;
```
Comment (pcwalton) [inline components/layout/flex.rs]: Can we factor this out into a method on block flows?

File: components/layout/flex.rs [hunk 8]
```diff
@@ -69,22 +72,192 @@ impl AxisSize {
...
+        let ref fragment = self.flow.as_block().fragment;
+        let adjustment = match mode {
+            Mode::Inline => {
+                match self.style.get_position().box_sizing {
+                    box_sizing::T::content_box =>
+                        fragment.border_padding.inline_start_end() + fragment.margin.inline_start_end(),
+                    box_sizing::T::border_box =>
+                        fragment.margin.inline_start_end()
+                }
+            },
```
Comment (pcwalton) [inline components/layout/flex.rs]: nit: no `,`

File: components/layout/flex.rs [hunk 9]
```diff
@@ -69,22 +72,192 @@ impl AxisSize {
...
+                        fragment.border_padding.block_start_end() + fragment.margin.block_start_end(),
+                    box_sizing::T::border_box =>
+                        fragment.margin.block_start_end()
+                }
+            }
+        };
+        max(self.min_size, min(self.base_size, self.max_size)) + adjustment
+    }
+
+    pub fn auto_margin_num(&self, mode: Mode) -> i32 {
```
Comment (pcwalton) [inline components/layout/flex.rs]: nit: How about `auto_margin_count`? Could possibly add a comment saying that this counts the number of margins in the provided direction.

File: components/layout/flex.rs [hunk 10]
```diff
@@ -262,6 +263,98 @@ impl FlexItem {
     }
 }
 
+/// A line in a flex container.
+// TODO(stshine): More fields are required to handle collapsed items and baseline alignment.
+#[derive(Debug)]
+struct FlexLine {
+    /// Range of items belong to this line in 'self.items'.
+    pub range: Range<usize>,
+    /// Remainig free space of this line, items will grow or shrink based on it being positive or negative.
```
Comment (pcwalton) [inline components/layout/flex.rs]: typo: "Remainig" -> "Remaining"

File: components/layout/flex.rs [hunk 11]
```diff
@@ -262,6 +263,98 @@ impl FlexItem {
...
+    pub free_space: Au,
+    /// the number of auto margins of items.
```
Comment (pcwalton) [inline components/layout/flex.rs]: uber-nit: capitalize "the"

File: components/layout/flex.rs [hunk 12]
```diff
@@ -421,6 +421,42 @@ impl FlexFlow {
...
+    /// 'container_size' param does not need to be 'None'. A line has to contain at least one item;
+    /// (expect this) if the container can be multi-line the sum of outer main size of items should
+    /// be less than the container size; a line should be filled by items as much as possible.
+    /// After been collected in a line a item should have its main sizes initialized.
+    fn get_flex_line(&mut self, container_size: Au) -> Option<FlexLine> {
+        let start = if self.lines.len() == 0 {
+            0
+        } else {
+            self.lines[self.lines.len()-1].range.end
+        };
```
Comment (pcwalton) [inline components/layout/flex.rs]: Could be just `let start = self.lines.last().map(|line| line.range.end).unwrap_or(0);`

File: components/layout/flex.rs [hunk 13]
```diff
@@ -421,6 +421,42 @@ impl FlexFlow {
...
+        }
+        let mut end = start;
+        let mut total_line_size = Au(0);
+        let mut margin_count = 0;
+
+        let items = &mut self.items[start..];
+        for mut item in items.iter_mut() {
+            item.init_sizes(container_size, self.main_mode);
+            let outer_main_size = item.outer_main_size(self.main_mode);
+            if total_line_size + outer_main_size > container_size && end != start  && self.is_wrappable {
```
Comment (pcwalton) [inline components/layout/flex.rs]: nit: just one space before the `&&`

File: components/layout/flex.rs [hunk 14]
```diff
@@ -421,6 +421,42 @@ impl FlexFlow {
...
+        };
+        if start == self.items.len() {
+            return None;
+        }
+        let mut end = start;
+        let mut total_line_size = Au(0);
+        let mut margin_count = 0;
+
+        let items = &mut self.items[start..];
+        for mut item in items.iter_mut() {
```
Comment (pcwalton) [inline components/layout/flex.rs]: nit: could just be `for mut item in items`

File: components/layout/block.rs [hunk 15]
```diff
@@ -3057,6 +3077,30 @@ impl ISizeAndMarginsComputer for InlineBlockReplaced {
...
+    fn compute_used_inline_size(&self,
+                                block: &mut BlockFlow,
+                                shared_context: &SharedStyleContext,
+                                parent_flow_inline_size: Au) {
+        let container_block_size = block.explicit_block_containing_size(shared_context);
+        block.fragment.assign_replaced_inline_size_if_necessary(parent_flow_inline_size,
+                                                                container_block_size);
+    }
+
+    // Literally do nothing.
```
Comment (pcwalton) [inline components/layout/block.rs]: I would say something like "The inline size has already been set by our flex flow parent, so do nothing here."

File: components/layout/flex.rs [hunk 16]
```diff
@@ -581,30 +577,106 @@ impl FlexFlow {
...
-            base.block_container_explicit_block_size = block_container_explicit_block_size;
-            if !self.main_reverse {
-              base.position.start.i = inline_child_start;
-              inline_child_start = inline_child_start + even_content_inline_size;
+        let parent_container_size = self.block_flow.explicit_block_containing_size(_shared_context);
+        // https://drafts.csswg.org/css-ui-3/#box-sizing
+        let explicit_content_size = self
+                                    .block_flow
+                                    .explicit_block_size(parent_container_size)
+                                    .map(|x| if x < box_border { Au(0) } else { x - box_border });
```
Comment (pcwalton) [inline components/layout/flex.rs]: Could just be `max(x - box_border, Au(0))`

File: components/layout/flex.rs [hunk 17]
```diff
@@ -581,30 +577,106 @@ impl FlexFlow {
...
+                block.fragment.margin.inline_end = margin_inline_end;
+                block.fragment.border_box.start.i = margin_inline_start;
+                block.fragment.border_box.size.inline = item_inline_size;
+                if !self.main_reverse {
+                    block.base.position.start.i = cur_i;
+                } else {
+                    block.base.position.start.i =
+                        inline_start_content_edge *2 + content_inline_size
+                        - cur_i  - item_outer_size;
+                };
```
Comment (pcwalton) [inline components/layout/flex.rs]: Strange formatting. Perhaps:

    block.base.position.start.i = if !self.main_reverse {
        cur_i
    } else {
        inline_start_content_edge * 2 + content_inline_size - cur_i - item_outer_size
    };

File: components/layout/flex.rs [hunk 18]
```diff
@@ -628,63 +700,137 @@ impl FlexFlow {
...
+            box_sizing::T::border_box => self.block_flow.fragment.border_padding.block_start_end(),
+            box_sizing::T::content_box => Au(0),
         };
+        let parent_container_size =
+            self.block_flow.explicit_block_containing_size(layout_context.shared_context());
+        // https://drafts.csswg.org/css-ui-3/#box-sizing
+        let explicit_content_size = self
+                                    .block_flow
+                                    .explicit_block_size(parent_container_size)
+                                    .map(|x| if x < box_border { Au(0) } else { x - box_border });
```
Comment (pcwalton) [inline components/layout/flex.rs]: Could use `max()` here again

File: components/layout/flex.rs [hunk 19]
```diff
@@ -628,63 +700,137 @@ impl FlexFlow {
...
-        let mut position = self.block_flow.fragment.border_box;
-        position.size.block = block_size;
-        self.block_flow.fragment.border_box = position;
-        self.block_flow.base.position.size.block = block_size;
+            line_interval = match line_align {
+                align_content::T::space_between => {
+                    if line_count == 1 {
+                        Au(0)
+                    } else {
+                        free_space /(line_count - 1)
```
Comment (pcwalton) [inline components/layout/flex.rs]: uber-nit: space after `/`

File: components/layout/flex.rs [hunk 20]
```diff
@@ -628,63 +700,137 @@ impl FlexFlow {
...
+                        block.fragment.border_box.size.block = block.base.position.size.block;
+                        // FIXME(stshine): item with 'align-self: stretch' and auto cross size should act
+                        // as if it has a fixed cross size, all child blocks should resolve against it.
+                        // block.assign_block_size(layout_context);
+                    }
+                block.base.position.start.b = margin_block_start +
+                    if !self.cross_reverse {
+                        cur_b
+                    } else {
+                        self.block_flow.fragment.border_padding.block_start *2
```
Comment (pcwalton) [inline components/layout/flex.rs]: nit: space after `*`

**Comment by stshine** [comment]
I refactored the 'box-sizing: border box' calculation into a method of `Fragment`. Other refactoring may be easier after make the `start_end*` methods of `LogicalMargin` accept a direction parameter.

Any suggestion on the naming?

**Comment by pcwalton** [comment]
@stshine The naming you chose seems fine to me.

**Comment by stshine** [comment]
Can't understand why the last letter 'L' moved a bit in `flexbox_direction-row-reverse'. Maybe some intermittent error in text layout?
http://i.imgur.com/UcQy10x.png

**Comment by shinglyu** [comment]
Sometimes the font stuff in the CI is different from running locally. Can that test pass when you run locally?

**Comment by stshine** [comment]
@shinglyu I believe the last failure is an intermittent failure caused by #12552, and the flexbox tests should go correctly now, don't worry :laughing: 
Still not sure why that letter moves one pixel on mac specifically, maybe something related to subpixel.

**Comment by shinglyu** [comment]
Yeah probably, sometimes different platform rounds appunit differently and will cause the text to shift.

**Comment by shinglyu** [comment]
@stshine By the way, after you land this patch, would you mind opening some followup bugs for the missing features? If you can provide some detail about them, I believe some contributors will be able to jump in and help.

**Comment by stshine** [comment]
@shinglyu  Since this is just a basic flexbox implementation,  I have been collecting issues for a while, among which the most significant one is the column container support. Implementing this and `align-self:stretch` get complicated because 1. how a block determine its widths/height need to be dramatically changed, 2. this contradicts heavily with the parallel layout mechanism of servo. 

  But I need to sort all info carefully in my head and dive more to the layout system more to state those issues properly and with more details. I will reply to the main issue soon after this get merged.

**Comment by pcwalton** [comment]
This looks good to me. @stshine, let me know when you feel this is ready and I'll r=me. Sorry about taking so long here!

**Comment by stshine** [comment]
@pcwalton Sure it is! No problem, I know you are quite busy recently :) Thank you very much for taking time to review!

**Comment by stshine** [comment]
Some upstream changes break the margin collapsing of references, which is sad. (Maybe also #12696 and #12676)

**Comment by pcwalton** [comment]
🎉 Thanks for pushing this through, @stshine!

**Comment by stshine** [comment]
I feel satisfied and honored to be able to implement this for servo. Thank you @pcwalton again and thanks to everyone who helped!
